In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

In [ ]:
df = pd.read_csv("/content/Churn_Modelling.csv")

In [ ]:
x = df.iloc[:,3:13]
y = df.iloc[:,13]

In [ ]:
df

,RowNumber,CustomerId,Surname,CreditScore,Geography,Gender,Age,Tenure,Balance,NumOfProducts,HasCrCard,IsActiveMember,EstimatedSalary,Exited
0,1,15634602,Hargrave,619,France,Female,42,2,0.00,1,1,1,101348.88,1
1,2,15647311,Hill,608,Spain,Female,41,1,83807.86,1,0,1,112542.58,0
2,3,15619304,Onio,502,France,Female,42,8,159660.80,3,1,0,113931.57,1
3,4,15701354,Boni,699,France,Female,39,1,0.00,2,0,0,93826.63,0
4,5,15737888,Mitchell,850,Spain,Female,43,2,125510.82,1,1,1,79084.10,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
9995,9996,15606229,Obijiaku,771,France,Male,39,5,0.00,2,1,0,96270.64,0
9996,9997,15569892,Johnstone,516,France,Male,35,10,57369.61,1,1,1,101699.77,0
9997,9998,15584532,Liu,709,France,Female,36,7,0.00,1,0,1,42085.58,1
9998,9999,15682355,Sabbatini,772,Germany,Male,42,3,75075.31,2,1,0,92888.52,1


In [ ]:
df.head()

,RowNumber,CustomerId,Surname,CreditScore,Geography,Gender,Age,Tenure,Balance,NumOfProducts,HasCrCard,IsActiveMember,EstimatedSalary,Exited
0,1,15634602,Hargrave,619,France,Female,42,2,0.00,1,1,1,101348.88,1
1,2,15647311,Hill,608,Spain,Female,41,1,83807.86,1,0,1,112542.58,0
2,3,15619304,Onio,502,France,Female,42,8,159660.80,3,1,0,113931.57,1
3,4,15701354,Boni,699,France,Female,39,1,0.00,2,0,0,93826.63,0
4,5,15737888,Mitchell,850,Spain,Female,43,2,125510.82,1,1,1,79084.10,0


In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   RowNumber        10000 non-null  int64  
 1   CustomerId       10000 non-null  int64  
 2   Surname          10000 non-null  object 
 3   CreditScore      10000 non-null  int64  
 4   Geography        10000 non-null  object 
 5   Gender           10000 non-null  object 
 6   Age              10000 non-null  int64  
 7   Tenure           10000 non-null  int64  
 8   Balance          10000 non-null  float64
 9   NumOfProducts    10000 non-null  int64  
 10  HasCrCard        10000 non-null  int64  
 11  IsActiveMember   10000 non-null  int64  
 12  EstimatedSalary  10000 non-null  float64
 13  Exited           10000 non-null  int64  
dtypes: float64(2), int64(9), object(3)
memory usage: 1.1+ MB


In [ ]:
x

,CreditScore,Geography,Gender,Age,Tenure,Balance,NumOfProducts,HasCrCard,IsActiveMember,EstimatedSalary
0,619,France,Female,42,2,0.00,1,1,1,101348.88
1,608,Spain,Female,41,1,83807.86,1,0,1,112542.58
2,502,France,Female,42,8,159660.80,3,1,0,113931.57
3,699,France,Female,39,1,0.00,2,0,0,93826.63
4,850,Spain,Female,43,2,125510.82,1,1,1,79084.10
...,...,...,...,...,...,...,...,...,...,...
9995,771,France,Male,39,5,0.00,2,1,0,96270.64
9996,516,France,Male,35,10,57369.61,1,1,1,101699.77
9997,709,France,Female,36,7,0.00,1,0,1,42085.58
9998,772,Germany,Male,42,3,75075.31,2,1,0,92888.52


In [ ]:
from sklearn.preprocessing import LabelEncoder

labelencoder_x_1 = LabelEncoder()
x.iloc[:, 1] = labelencoder_x_1.fit_transform(x.iloc[:, 1])

labelencoder_x_2 = LabelEncoder()
x.iloc[:, 2] = labelencoder_x_2.fit_transform(x.iloc[:, 2])

In [ ]:
x

,CreditScore,Geography,Gender,Age,Tenure,Balance,NumOfProducts,HasCrCard,IsActiveMember,EstimatedSalary
0,619,0,0,42,2,0.00,1,1,1,101348.88
1,608,2,0,41,1,83807.86,1,0,1,112542.58
2,502,0,0,42,8,159660.80,3,1,0,113931.57
3,699,0,0,39,1,0.00,2,0,0,93826.63
4,850,2,0,43,2,125510.82,1,1,1,79084.10
...,...,...,...,...,...,...,...,...,...,...
9995,771,0,1,39,5,0.00,2,1,0,96270.64
9996,516,0,1,35,10,57369.61,1,1,1,101699.77
9997,709,0,0,36,7,0.00,1,0,1,42085.58
9998,772,1,1,42,3,75075.31,2,1,0,92888.52


In [ ]:
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size = 0.2, random_state = 0)

In [ ]:
len(df)

10000

In [ ]:
len(x_test)

2000

In [ ]:
from sklearn.preprocessing import StandardScaler

In [ ]:
ss = StandardScaler()
x_train = ss.fit_transform(x_train)
x_test = ss.transform(x_test)

In [ ]:
from sklearn.tree import DecisionTreeClassifier

In [ ]:
classifier = DecisionTreeClassifier()
classifier.fit(x_train, y_train)
pred = classifier.predict(x_test)

In [ ]:
from sklearn.metrics import confusion_matrix

In [ ]:
conM = confusion_matrix(y_test, pred)
print(conM)

[[1361  234]
 [ 186  219]]


In [ ]:
from sklearn.metrics import accuracy_score, classification_report

In [ ]:
print(accuracy_score(y_test, pred))
print(classification_report(y_test, pred))

0.79
              precision    recall  f1-score   support

           0       0.88      0.85      0.87      1595
           1       0.48      0.54      0.51       405

    accuracy                           0.79      2000
   macro avg       0.68      0.70      0.69      2000
weighted avg       0.80      0.79      0.79      2000



In [ ]:
from sklearn.ensemble import RandomForestClassifier

In [ ]:
rfc = RandomForestClassifier()
rfc.fit(x_train, y_train)
pred = rfc.predict(x_test)

In [ ]:
from sklearn.metrics import confusion_matrix

In [ ]:
conM = confusion_matrix(y_test, pred)
print(conM)

[[1522   73]
 [ 202  203]]


In [ ]:
from sklearn.metrics import accuracy_score, classification_report

In [ ]:
print(accuracy_score(y_test, pred))
print(classification_report(y_test, pred))

0.8625
              precision    recall  f1-score   support

           0       0.88      0.95      0.92      1595
           1       0.74      0.50      0.60       405

    accuracy                           0.86      2000
   macro avg       0.81      0.73      0.76      2000
weighted avg       0.85      0.86      0.85      2000



In [ ]:
#Begining with ANN from here
#first import keras

#pip install keras

In [ ]:
#pip install tensorflow

In [ ]:
import keras
from keras.models import Sequential
from keras.layers import Dense

In [ ]:
#Initializing ANN
# 5 Hidden Layers

classifier = Sequential()

classifier.add(Dense(units=6,kernel_initializer='uniform',activation='relu',input_dim=10))
classifier.add(Dense(units=6,kernel_initializer='uniform',activation='relu'))
classifier.add(Dense(units=6,kernel_initializer='uniform',activation='relu'))
classifier.add(Dense(units=6,kernel_initializer='uniform',activation='relu'))
classifier.add(Dense(units=6,kernel_initializer='uniform',activation='relu'))
classifier.add(Dense(units=6,kernel_initializer='uniform',activation='relu'))
classifier.add(Dense(units=1,kernel_initializer='uniform',activation='sigmoid'))

/usr/local/lib/python3.12/dist-packages/keras/src/layers/core/dense.py:106: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


In [ ]:
len(x_train)

8000

In [ ]:
#compiling the ANN
classifier.compile(optimizer='adam',loss='binary_crossentropy',metrics=['accuracy'])
y_train.iloc[0]

np.int64(0)

In [ ]:
classifier.fit(x_train, y_train, batch_size=10,epochs=100)

Epoch 1/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 6s 2ms/step - accuracy: 0.7960 - loss: 0.5011
Epoch 2/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 3s 4ms/step - accuracy: 0.7960 - loss: 0.4366
Epoch 3/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 4s 2ms/step - accuracy: 0.7960 - loss: 0.4320
Epoch 4/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 3s 2ms/step - accuracy: 0.8056 - loss: 0.4263
Epoch 5/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8226 - loss: 0.4231
Epoch 6/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8275 - loss: 0.4207
Epoch 7/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 3s 4ms/step - accuracy: 0.8285 - loss: 0.4194
Epoch 8/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 4s 2ms/step - accuracy: 0.8299 - loss: 0.4176
Epoch 9/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - accuracy: 0.8310 - loss: 0.4162
Epoch 10/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8326 - loss: 0.4142
Epoch 11/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8341 - loss: 0.4129
Epoch 12/100
800/800 ━━━━━━━━━━━━━━━━━━━━

In [ ]:
y_pred_ann=classifier.predict(x_test)
y_pred_ann=(y_pred_ann>0.5)
cm3=confusion_matrix(y_test,y_pred_ann)
print(cm3)

63/63 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step
[[1553   42]
 [ 281  124]]


In [ ]:
from sklearn.metrics import accuracy_score
print(accuracy_score(y_test,y_pred_ann))

0.8385


In [ ]:
#Initializing ANN
# 3 Hidden Layers

classifier = Sequential()

classifier.add(Dense(units=6,kernel_initializer='uniform',activation='relu',input_dim=10))
classifier.add(Dense(units=6,kernel_initializer='uniform',activation='relu'))
classifier.add(Dense(units=6,kernel_initializer='uniform',activation='relu'))
classifier.add(Dense(units=6,kernel_initializer='uniform',activation='relu'))
classifier.add(Dense(units=1,kernel_initializer='uniform',activation='sigmoid'))

/usr/local/lib/python3.12/dist-packages/keras/src/layers/core/dense.py:106: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


In [ ]:
len(x_train)

8000

In [ ]:
#compiling the ANN
classifier.compile(optimizer='adam',loss='binary_crossentropy',metrics=['accuracy'])
y_train.iloc[0]

np.int64(0)

In [ ]:
classifier.fit(x_train, y_train, batch_size=10,epochs=100)

Epoch 1/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 4s 2ms/step - accuracy: 0.7951 - loss: 0.4988
Epoch 2/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 3s 2ms/step - accuracy: 0.7960 - loss: 0.4362
Epoch 3/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.7960 - loss: 0.4320
Epoch 4/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - accuracy: 0.7970 - loss: 0.4274
Epoch 5/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - accuracy: 0.8217 - loss: 0.4229
Epoch 6/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8261 - loss: 0.4203
Epoch 7/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8275 - loss: 0.4182
Epoch 8/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8288 - loss: 0.4171
Epoch 9/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8322 - loss: 0.4150
Epoch 10/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8313 - loss: 0.4137
Epoch 11/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 3s 4ms/step - accuracy: 0.8324 - loss: 0.4128
Epoch 12/100
800/800 ━━━━━━━━━━━━━━━━━━━━

In [ ]:
y_pred_ann=classifier.predict(x_test)
y_pred_ann=(y_pred_ann>0.5)
cm3=confusion_matrix(y_test,y_pred_ann)
print(cm3)

63/63 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step
[[1536   59]
 [ 249  156]]


In [ ]:
from sklearn.metrics import accuracy_score
print(accuracy_score(y_test,y_pred_ann))

0.846


In [ ]:
#Initializing ANN
# 2 Hidden Layers

classifier = Sequential()

classifier.add(Dense(units=6,kernel_initializer='uniform',activation='relu',input_dim=10))
classifier.add(Dense(units=6,kernel_initializer='uniform',activation='relu'))
classifier.add(Dense(units=6,kernel_initializer='uniform',activation='relu'))
classifier.add(Dense(units=1,kernel_initializer='uniform',activation='sigmoid'))

/usr/local/lib/python3.12/dist-packages/keras/src/layers/core/dense.py:106: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


In [ ]:
len(x_train)

8000

In [ ]:
#compiling the ANN
classifier.compile(optimizer='adam',loss='binary_crossentropy',metrics=['accuracy'])
y_train.iloc[0]

np.int64(0)

In [ ]:
classifier.fit(x_train, y_train, batch_size=10,epochs=100)

Epoch 1/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 3s 2ms/step - accuracy: 0.7958 - loss: 0.4879
Epoch 2/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8151 - loss: 0.4258
Epoch 3/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8299 - loss: 0.4179
Epoch 4/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 3s 3ms/step - accuracy: 0.8340 - loss: 0.4118
Epoch 5/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8350 - loss: 0.4093
Epoch 6/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8342 - loss: 0.4080
Epoch 7/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8344 - loss: 0.4068
Epoch 8/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8342 - loss: 0.4066
Epoch 9/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8339 - loss: 0.4061
Epoch 10/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8350 - loss: 0.4056
Epoch 11/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - accuracy: 0.8349 - loss: 0.4049
Epoch 12/100
800/800 ━━━━━━━━━━━━━━━━━━━━

In [ ]:
y_pred_ann=classifier.predict(x_test)
y_pred_ann=(y_pred_ann>0.5)
cm3=confusion_matrix(y_test,y_pred_ann)
print(cm3)

63/63 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step
[[1540   55]
 [ 264  141]]


In [ ]:
from sklearn.metrics import accuracy_score
print(accuracy_score(y_test,y_pred_ann))

0.8405


In [ ]:
#Initializing ANN
# 2 Hidden Layers

classifier = Sequential()

classifier.add(Dense(units=6,kernel_initializer='uniform',activation='relu',input_dim=10))
classifier.add(Dense(units=6,kernel_initializer='uniform',activation='relu'))
classifier.add(Dense(units=1,kernel_initializer='uniform',activation='sigmoid'))

/usr/local/lib/python3.12/dist-packages/keras/src/layers/core/dense.py:106: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


In [ ]:
len(x_train)

8000

In [ ]:
#compiling the ANN
classifier.compile(optimizer='adam',loss='binary_crossentropy',metrics=['accuracy'])
y_train.iloc[0]

np.int64(0)

In [ ]:
classifier.fit(x_train, y_train, batch_size=10,epochs=100)

Epoch 1/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 3s 2ms/step - accuracy: 0.7954 - loss: 0.4931
Epoch 2/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 3s 2ms/step - accuracy: 0.7960 - loss: 0.4342
Epoch 3/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.7960 - loss: 0.4292
Epoch 4/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8035 - loss: 0.4249
Epoch 5/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 3s 3ms/step - accuracy: 0.8253 - loss: 0.4219
Epoch 6/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8270 - loss: 0.4200
Epoch 7/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8286 - loss: 0.4177
Epoch 8/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8315 - loss: 0.4158
Epoch 9/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8314 - loss: 0.4148
Epoch 10/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.8336 - loss: 0.4135
Epoch 11/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8332 - loss: 0.4128
Epoch 12/100
800/800 ━━━━━━━━━━━━━━━━━━━━

In [ ]:
y_pred_ann=classifier.predict(x_test)
y_pred_ann=(y_pred_ann>0.5)
cm3=confusion_matrix(y_test,y_pred_ann)
print(cm3)

63/63 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step
[[1545   50]
 [ 263  142]]


In [ ]:
from sklearn.metrics import accuracy_score
print(accuracy_score(y_test,y_pred_ann))

0.8435


In [ ]:
# TASK ONE COMPLETED

In [ ]:
# BEGINING OF TASK - 2

In [ ]:
#Initializing ANN
# 2 Hidden Layers, changing units to 4

classifier = Sequential()

classifier.add(Dense(units=6,kernel_initializer='uniform',activation='relu',input_dim=10))
classifier.add(Dense(units=4,kernel_initializer='uniform',activation='relu'))
classifier.add(Dense(units=4,kernel_initializer='uniform',activation='relu'))
classifier.add(Dense(units=1,kernel_initializer='uniform',activation='sigmoid'))

/usr/local/lib/python3.12/dist-packages/keras/src/layers/core/dense.py:106: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


In [ ]:
#compiling the ANN
classifier.compile(optimizer='adam',loss='binary_crossentropy',metrics=['accuracy'])
y_train.iloc[0]

np.int64(0)

In [ ]:
classifier.fit(x_train, y_train, batch_size=10,epochs=100)

Epoch 1/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 4s 3ms/step - accuracy: 0.7950 - loss: 0.5050
Epoch 2/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.7960 - loss: 0.4305
Epoch 3/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.7960 - loss: 0.4279
Epoch 4/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.7968 - loss: 0.4254
Epoch 5/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8226 - loss: 0.4235
Epoch 6/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8279 - loss: 0.4208
Epoch 7/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - accuracy: 0.8276 - loss: 0.4182
Epoch 8/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - accuracy: 0.8295 - loss: 0.4156
Epoch 9/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8305 - loss: 0.4125
Epoch 10/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8282 - loss: 0.4105
Epoch 11/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8295 - loss: 0.4073
Epoch 12/100
800/800 ━━━━━━━━━━━━━━━━━━━━

In [ ]:
y_pred_ann=classifier.predict(x_test)
y_pred_ann=(y_pred_ann>0.5)
cm3=confusion_matrix(y_test,y_pred_ann)
print(cm3)

63/63 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step
[[1522   73]
 [ 200  205]]


In [ ]:
from sklearn.metrics import accuracy_score
print(accuracy_score(y_test,y_pred_ann))

0.8635


In [ ]:
#Initializing ANN
# 2 Hidden Layers, changing units to 8

classifier = Sequential()

classifier.add(Dense(units=6,kernel_initializer='uniform',activation='relu',input_dim=10))
classifier.add(Dense(units=8,kernel_initializer='uniform',activation='relu'))
classifier.add(Dense(units=8,kernel_initializer='uniform',activation='relu'))
classifier.add(Dense(units=1,kernel_initializer='uniform',activation='sigmoid'))

/usr/local/lib/python3.12/dist-packages/keras/src/layers/core/dense.py:106: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


In [ ]:
#compiling the ANN
classifier.compile(optimizer='adam',loss='binary_crossentropy',metrics=['accuracy'])
y_train.iloc[0]

np.int64(0)

In [ ]:
classifier.fit(x_train, y_train, batch_size=10,epochs=100)

Epoch 1/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 3s 2ms/step - accuracy: 0.7960 - loss: 0.4820
Epoch 2/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.7960 - loss: 0.4302
Epoch 3/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8139 - loss: 0.4232
Epoch 4/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 3s 3ms/step - accuracy: 0.8263 - loss: 0.4195
Epoch 5/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - accuracy: 0.8316 - loss: 0.4164
Epoch 6/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 3s 4ms/step - accuracy: 0.8321 - loss: 0.4140
Epoch 7/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 4s 2ms/step - accuracy: 0.8334 - loss: 0.4125
Epoch 8/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - accuracy: 0.8340 - loss: 0.4103
Epoch 9/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - accuracy: 0.8331 - loss: 0.4093
Epoch 10/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8334 - loss: 0.4079
Epoch 11/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8325 - loss: 0.4070
Epoch 12/100
800/800 ━━━━━━━━━━━━━━━━━━━━

In [ ]:
y_pred_ann=classifier.predict(x_test)
y_pred_ann=(y_pred_ann>0.5)
cm3=confusion_matrix(y_test,y_pred_ann)
print(cm3)

63/63 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step
[[1504   91]
 [ 198  207]]


In [ ]:
from sklearn.metrics import accuracy_score
print(accuracy_score(y_test,y_pred_ann))

0.8555


In [ ]:
#Initializing ANN
# 2 Hidden Layers, changing units to 16

classifier = Sequential()

classifier.add(Dense(units=6,kernel_initializer='uniform',activation='relu',input_dim=10))
classifier.add(Dense(units=16,kernel_initializer='uniform',activation='relu'))
classifier.add(Dense(units=16,kernel_initializer='uniform',activation='relu'))
classifier.add(Dense(units=1,kernel_initializer='uniform',activation='sigmoid'))

/usr/local/lib/python3.12/dist-packages/keras/src/layers/core/dense.py:106: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


In [ ]:
#compiling the ANN
classifier.compile(optimizer='adam',loss='binary_crossentropy',metrics=['accuracy'])
y_train.iloc[0]

np.int64(0)

In [ ]:
classifier.fit(x_train, y_train, batch_size=10,epochs=100)

Epoch 1/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 3s 2ms/step - accuracy: 0.7956 - loss: 0.4721
Epoch 2/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - accuracy: 0.7961 - loss: 0.4306
Epoch 3/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - accuracy: 0.8183 - loss: 0.4243
Epoch 4/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8267 - loss: 0.4211
Epoch 5/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8274 - loss: 0.4194
Epoch 6/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8311 - loss: 0.4177
Epoch 7/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8311 - loss: 0.4161
Epoch 8/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8313 - loss: 0.4151
Epoch 9/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 3s 3ms/step - accuracy: 0.8324 - loss: 0.4148
Epoch 10/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - accuracy: 0.8314 - loss: 0.4131
Epoch 11/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8330 - loss: 0.4129
Epoch 12/100
800/800 ━━━━━━━━━━━━━━━━━━━━

In [ ]:
y_pred_ann=classifier.predict(x_test)
y_pred_ann=(y_pred_ann>0.5)
cm3=confusion_matrix(y_test,y_pred_ann)
print(cm3)

63/63 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step
[[1539   56]
 [ 252  153]]


In [ ]:
from sklearn.metrics import accuracy_score
print(accuracy_score(y_test,y_pred_ann))

0.846


In [ ]:
#Initializing ANN
# 2 Hidden Layers, changing units to 32

classifier = Sequential()

classifier.add(Dense(units=6,kernel_initializer='uniform',activation='relu',input_dim=10))
classifier.add(Dense(units=32,kernel_initializer='uniform',activation='relu'))
classifier.add(Dense(units=32,kernel_initializer='uniform',activation='relu'))
classifier.add(Dense(units=1,kernel_initializer='uniform',activation='sigmoid'))

/usr/local/lib/python3.12/dist-packages/keras/src/layers/core/dense.py:106: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


In [ ]:
#compiling the ANN
classifier.compile(optimizer='adam',loss='binary_crossentropy',metrics=['accuracy'])
y_train.iloc[0]

np.int64(0)

In [ ]:
classifier.fit(x_train, y_train, batch_size=10,epochs=100)

Epoch 1/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 7s 5ms/step - accuracy: 0.7945 - loss: 0.4711
Epoch 2/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 3s 4ms/step - accuracy: 0.8069 - loss: 0.4272
Epoch 3/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - accuracy: 0.8242 - loss: 0.4230
Epoch 4/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8300 - loss: 0.4195
Epoch 5/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8307 - loss: 0.4166
Epoch 6/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 3s 3ms/step - accuracy: 0.8320 - loss: 0.4154
Epoch 7/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8329 - loss: 0.4129
Epoch 8/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8339 - loss: 0.4123
Epoch 9/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8332 - loss: 0.4109
Epoch 10/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8328 - loss: 0.4100
Epoch 11/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8345 - loss: 0.4086
Epoch 12/100
800/800 ━━━━━━━━━━━━━━━━━━━━

In [ ]:
y_pred_ann=classifier.predict(x_test)
y_pred_ann=(y_pred_ann>0.5)
cm3=confusion_matrix(y_test,y_pred_ann)
print(cm3)

63/63 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step
[[1517   78]
 [ 204  201]]


In [ ]:
from sklearn.metrics import accuracy_score
print(accuracy_score(y_test,y_pred_ann))

0.859


In [ ]:
#Initializing ANN
# 2 Hidden Layers, changing units to 64

classifier = Sequential()

classifier.add(Dense(units=6,kernel_initializer='uniform',activation='relu',input_dim=10))
classifier.add(Dense(units=64,kernel_initializer='uniform',activation='relu'))
classifier.add(Dense(units=64,kernel_initializer='uniform',activation='relu'))
classifier.add(Dense(units=1,kernel_initializer='uniform',activation='sigmoid'))

In [ ]:
#compiling the ANN
classifier.compile(optimizer='adam',loss='binary_crossentropy',metrics=['accuracy'])
y_train.iloc[0]

np.int64(0)

In [ ]:
classifier.fit(x_train, y_train, batch_size=10,epochs=100)

Epoch 1/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 6s 3ms/step - accuracy: 0.7952 - loss: 0.4625
Epoch 2/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 3s 3ms/step - accuracy: 0.8092 - loss: 0.4272
Epoch 3/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 3s 4ms/step - accuracy: 0.8267 - loss: 0.4217
Epoch 4/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 6s 7ms/step - accuracy: 0.8294 - loss: 0.4171
Epoch 5/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 5s 7ms/step - accuracy: 0.8295 - loss: 0.4146
Epoch 6/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 6s 8ms/step - accuracy: 0.8345 - loss: 0.4124
Epoch 7/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 3s 4ms/step - accuracy: 0.8332 - loss: 0.4080
Epoch 8/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 5s 3ms/step - accuracy: 0.8330 - loss: 0.4075
Epoch 9/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 4s 2ms/step - accuracy: 0.8330 - loss: 0.4065
Epoch 10/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8335 - loss: 0.4049
Epoch 11/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8314 - loss: 0.4037
Epoch 12/100
800/800 ━━━━━━━━━━━━━━━━━━━━

In [ ]:
y_pred_ann=classifier.predict(x_test)
y_pred_ann=(y_pred_ann>0.5)
cm3=confusion_matrix(y_test,y_pred_ann)
print(cm3)

63/63 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step
[[1501   94]
 [ 197  208]]


In [ ]:
from sklearn.metrics import accuracy_score
print(accuracy_score(y_test,y_pred_ann))

0.8545


Create a comparison table.

Hidden Neurons  Accuracy  LossTraining Time
4
8
16

32
64

### ANN Model Comparison

| Hidden Layers | Neurons | Test Accuracy | Last Epoch Training Loss | Estimated Training Time (s) |
|---------------|---------|---------------|--------------------------|-----------------------------|
| 2             | 4       | 0.8635        | 0.3408                   | ~300                        |
| 2             | 8       | 0.8555        | 0.3521                   | ~300                        |
| 2             | 16      | 0.8545        | 0.3360                   | ~250                        |
| 2             | 32      | 0.859         | 0.3342                   | ~300                        |
| 2             | 64      | 0.8545        | 0.3360                   | ~350                        |

In [ ]:
# TASK 2 COMPLETED

In [ ]:
#TASK 3 - BEGIN

In [ ]:
#Initializing ANN
# 2 Hidden Layers , activation function = sigmoid

classifier = Sequential()

classifier.add(Dense(units=6,kernel_initializer='uniform',activation='relu',input_dim=10))
classifier.add(Dense(units=6,kernel_initializer='uniform',activation='sigmoid'))
classifier.add(Dense(units=6,kernel_initializer='uniform',activation='sigmoid'))
classifier.add(Dense(units=1,kernel_initializer='uniform',activation='sigmoid'))

/usr/local/lib/python3.12/dist-packages/keras/src/layers/core/dense.py:106: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


In [ ]:
#compiling the ANN
classifier.compile(optimizer='adam',loss='binary_crossentropy',metrics=['accuracy'])
y_train.iloc[0]

np.int64(0)

In [ ]:
classifier.fit(x_train, y_train, batch_size=10,epochs=100)

Epoch 1/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 5s 3ms/step - accuracy: 0.7943 - loss: 0.5290
Epoch 2/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 3s 4ms/step - accuracy: 0.7960 - loss: 0.4920
Epoch 3/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 4s 4ms/step - accuracy: 0.7960 - loss: 0.4822
Epoch 4/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.7960 - loss: 0.4701
Epoch 5/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.7960 - loss: 0.4586
Epoch 6/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.7960 - loss: 0.4498
Epoch 7/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.7960 - loss: 0.4440
Epoch 8/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.7960 - loss: 0.4406
Epoch 9/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 3s 3ms/step - accuracy: 0.7960 - loss: 0.4379
Epoch 10/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 3s 3ms/step - accuracy: 0.7960 - loss: 0.4361
Epoch 11/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.7960 - loss: 0.4345
Epoch 12/100
800/800 ━━━━━━━━━━━━━━━━━━━━

In [ ]:
y_pred_ann=classifier.predict(x_test)
y_pred_ann=(y_pred_ann>0.5)
cm3=confusion_matrix(y_test,y_pred_ann)
print(cm3)

63/63 ━━━━━━━━━━━━━━━━━━━━ 1s 10ms/step
[[1525   70]
 [ 245  160]]


In [ ]:
from sklearn.metrics import accuracy_score
print(accuracy_score(y_test,y_pred_ann))

0.8425


In [ ]:
#Initializing ANN
# 2 Hidden Layers , activation function = tanh

classifier = Sequential()

classifier.add(Dense(units=6,kernel_initializer='uniform',activation='relu',input_dim=10))
classifier.add(Dense(units=6,kernel_initializer='uniform',activation='tanh'))
classifier.add(Dense(units=6,kernel_initializer='uniform',activation='tanh'))
classifier.add(Dense(units=1,kernel_initializer='uniform',activation='sigmoid'))

/usr/local/lib/python3.12/dist-packages/keras/src/layers/core/dense.py:106: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


In [ ]:
#compiling the ANN
classifier.compile(optimizer='adam',loss='binary_crossentropy',metrics=['accuracy'])
y_train.iloc[0]

np.int64(0)

In [ ]:
classifier.fit(x_train, y_train, batch_size=10,epochs=100)

Epoch 1/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 6s 5ms/step - accuracy: 0.7955 - loss: 0.4905
Epoch 2/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 4s 5ms/step - accuracy: 0.8260 - loss: 0.4217
Epoch 3/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 4s 3ms/step - accuracy: 0.8335 - loss: 0.4130
Epoch 4/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 3s 4ms/step - accuracy: 0.8330 - loss: 0.4096
Epoch 5/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 7s 6ms/step - accuracy: 0.8339 - loss: 0.4077
Epoch 6/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 4s 5ms/step - accuracy: 0.8346 - loss: 0.4059
Epoch 7/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 5s 5ms/step - accuracy: 0.8349 - loss: 0.4044
Epoch 8/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 5s 6ms/step - accuracy: 0.8341 - loss: 0.4033
Epoch 9/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 3s 4ms/step - accuracy: 0.8330 - loss: 0.4033
Epoch 10/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 3s 3ms/step - accuracy: 0.8328 - loss: 0.4027
Epoch 11/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 3s 3ms/step - accuracy: 0.8332 - loss: 0.4024
Epoch 12/100
800/800 ━━━━━━━━━━━━━━━━━━━━

In [ ]:
y_pred_ann=classifier.predict(x_test)
y_pred_ann=(y_pred_ann>0.5)
cm3=confusion_matrix(y_test,y_pred_ann)
print(cm3)

63/63 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step
[[1558   37]
 [ 281  124]]


In [ ]:
from sklearn.metrics import accuracy_score
print(accuracy_score(y_test,y_pred_ann))

0.841


In [ ]:
#Initializing ANN
# 2 Hidden Layers , activation function = elu

classifier = Sequential()

classifier.add(Dense(units=6,kernel_initializer='uniform',activation='relu',input_dim=10))
classifier.add(Dense(units=6,kernel_initializer='uniform',activation='elu'))
classifier.add(Dense(units=6,kernel_initializer='uniform',activation='elu'))
classifier.add(Dense(units=1,kernel_initializer='uniform',activation='sigmoid'))

/usr/local/lib/python3.12/dist-packages/keras/src/layers/core/dense.py:106: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


In [ ]:
#compiling the ANN
classifier.compile(optimizer='adam',loss='binary_crossentropy',metrics=['accuracy'])
y_train.iloc[0]

np.int64(0)

In [ ]:
classifier.fit(x_train, y_train, batch_size=10,epochs=100)

Epoch 1/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 5s 3ms/step - accuracy: 0.7956 - loss: 0.4766
Epoch 2/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 7s 9ms/step - accuracy: 0.7960 - loss: 0.4351
Epoch 3/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 4s 5ms/step - accuracy: 0.7993 - loss: 0.4304
Epoch 4/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 3s 4ms/step - accuracy: 0.8270 - loss: 0.4220
Epoch 5/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 6s 5ms/step - accuracy: 0.8324 - loss: 0.4142
Epoch 6/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 3s 3ms/step - accuracy: 0.8339 - loss: 0.4100
Epoch 7/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 3s 3ms/step - accuracy: 0.8347 - loss: 0.4070
Epoch 8/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - accuracy: 0.8349 - loss: 0.4053
Epoch 9/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - accuracy: 0.8351 - loss: 0.4042
Epoch 10/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 4s 5ms/step - accuracy: 0.8351 - loss: 0.4025
Epoch 11/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8354 - loss: 0.4018
Epoch 12/100
800/800 ━━━━━━━━━━━━━━━━━━━━

In [ ]:
y_pred_ann=classifier.predict(x_test)
y_pred_ann=(y_pred_ann>0.5)
cm3=confusion_matrix(y_test,y_pred_ann)
print(cm3)

63/63 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step
[[1554   41]
 [ 267  138]]


In [ ]:
from sklearn.metrics import accuracy_score
print(accuracy_score(y_test,y_pred_ann))

0.846


In [ ]:
#Initializing ANN
# 2 Hidden Layers , activation function = leakyrelu

classifier = Sequential()

classifier.add(Dense(units=6,kernel_initializer='uniform',activation='relu',input_dim=10))
classifier.add(Dense(units=6,kernel_initializer='uniform',activation='leaky_relu'))
classifier.add(Dense(units=6,kernel_initializer='uniform',activation='leaky_relu'))
classifier.add(Dense(units=1,kernel_initializer='uniform',activation='sigmoid'))

In [ ]:
#compiling the ANN
classifier.compile(optimizer='adam',loss='binary_crossentropy',metrics=['accuracy'])
y_train.iloc[0]

np.int64(0)

In [ ]:
classifier.fit(x_train, y_train, batch_size=10,epochs=100)

Epoch 1/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 3s 2ms/step - accuracy: 0.7960 - loss: 0.4809
Epoch 2/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.7960 - loss: 0.4356
Epoch 3/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - accuracy: 0.7960 - loss: 0.4291
Epoch 4/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8255 - loss: 0.4214
Epoch 5/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8330 - loss: 0.4133
Epoch 6/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8344 - loss: 0.4097
Epoch 7/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8336 - loss: 0.4081
Epoch 8/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8356 - loss: 0.4070
Epoch 9/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8342 - loss: 0.4057
Epoch 10/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8355 - loss: 0.4042
Epoch 11/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - accuracy: 0.8365 - loss: 0.4038
Epoch 12/100
800/800 ━━━━━━━━━━━━━━━━━━━━

In [ ]:
y_pred_ann=classifier.predict(x_test)
y_pred_ann=(y_pred_ann>0.5)
cm3=confusion_matrix(y_test,y_pred_ann)
print(cm3)

63/63 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step
[[1544   51]
 [ 268  137]]


In [ ]:
from sklearn.metrics import accuracy_score
print(accuracy_score(y_test,y_pred_ann))

0.8405


In [ ]:
train_acc = history.history['accuracy'][-1]
loss = history.history['loss'][-1]

print("Training Accuracy:", train_acc)
print("Loss:", loss)

Training Accuracy: 0.8707500100135803
Loss: 0.3180745244026184


Compare

Accuracy
Loss
Epochs to Converge

### Simplified Model Comparison

| Model Type                | Test Accuracy | Last Epoch Training Loss | Total Epochs Trained |
|---------------------------|---------------|--------------------------|----------------------|
| Decision Tree Classifier  | 0.79          | N/A                      | N/A                  |
| Random Forest Classifier  | 0.8625        | N/A                      | N/A                  |
| ANN (5 Hidden Layers)     | 0.8385        | 0.3989                   | 100                  |
| ANN (3 Hidden Layers)     | 0.846         | 0.3972                   | 100                  |
| ANN (2 Hidden Layers)     | 0.8435        | 0.3894                   | 100                  |
| ANN (2 Hidden Layers, 4 units) | 0.8635        | 0.3408                   | 100                  |
| ANN (2 Hidden Layers, 8 units) | 0.8555        | 0.3521                   | 100                  |
| ANN (2 Hidden Layers, 16 units) | 0.846         | 0.3967                   | 100                  |
| ANN (2 Hidden Layers, 32 units) | 0.859         | 0.3342                   | 100                  |
| ANN (2 Hidden Layers, 64 units) | 0.8545        | 0.3360                   | 100                  |
| ANN (2 Hidden Layers, Sigmoid) | 0.8425        | 0.3942                   | 100                  |
| ANN (2 Hidden Layers, Tanh)    | 0.841         | 0.3926                   | 100                  |
| ANN (2 Hidden Layers, ELU)     | 0.846         | 0.3922                   | 100                  |
| ANN (2 Hidden Layers, Leaky ReLU) | 0.8405        | 0.3935                   | 100                  |

### ANN Model Comparison by Activation Function

| Model Type                      | Test Accuracy | Last Epoch Training Loss | Total Epochs Trained |
|---------------------------------|---------------|--------------------------|----------------------|
| ANN (2 Hidden Layers, Sigmoid)  | 0.8425        | 0.3942                   | 100                  |
| ANN (2 Hidden Layers, Tanh)     | 0.841         | 0.3926                   | 100                  |
| ANN (2 Hidden Layers, ELU)      | 0.846         | 0.3922                   | 100                  |
| ANN (2 Hidden Layers, Leaky ReLU) | 0.8405        | 0.3935                   | 100                  |

In [ ]:
#TASK 3 COMPLETED!

In [ ]:
#TASK - 4 BEGIN

In [174]:
#Initializing ANN
#kernel_initializer='he_uniform'

classifier = Sequential()

classifier.add(Dense(units=6, kernel_initializer='uniform', activation='relu', input_dim=10))
classifier.add(Dense(units=6, kernel_initializer='he_uniform', activation='relu'))
classifier.add(Dense(units=6, kernel_initializer='he_uniform', activation='relu'))
classifier.add(Dense(units=1, kernel_initializer='uniform', activation='sigmoid'))

/usr/local/lib/python3.12/dist-packages/keras/src/layers/core/dense.py:106: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


In [175]:
#compiling the ANN
classifier.compile(optimizer='adam',loss='binary_crossentropy',metrics=['accuracy'])
y_train.iloc[0]

np.int64(0)

In [176]:
classifier.fit(x_train, y_train, batch_size=10,epochs=100)

Epoch 1/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 3s 3ms/step - accuracy: 0.7960 - loss: 0.4658
Epoch 2/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - accuracy: 0.7984 - loss: 0.4180
Epoch 3/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8227 - loss: 0.4025
Epoch 4/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8338 - loss: 0.3902
Epoch 5/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8415 - loss: 0.3802
Epoch 6/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8443 - loss: 0.3757
Epoch 7/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8468 - loss: 0.3716
Epoch 8/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8469 - loss: 0.3683
Epoch 9/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 3s 3ms/step - accuracy: 0.8478 - loss: 0.3668
Epoch 10/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8506 - loss: 0.3645
Epoch 11/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8509 - loss: 0.3634
Epoch 12/100
800/800 ━━━━━━━━━━━━━━━━━━━━

In [177]:
y_pred_ann=classifier.predict(x_test)
y_pred_ann=(y_pred_ann>0.5)
cm3=confusion_matrix(y_test,y_pred_ann)
print(cm3)

63/63 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step
[[1535   60]
 [ 216  189]]


In [178]:
from sklearn.metrics import accuracy_score
print(accuracy_score(y_test,y_pred_ann))

0.862


In [179]:
#Initializing ANN
#kernel_initializer = he_normal

classifier = Sequential()

classifier.add(Dense(units=6, kernel_initializer='uniform', activation='relu', input_dim=10))
classifier.add(Dense(units=6, kernel_initializer='he_normal', activation='relu'))
classifier.add(Dense(units=6, kernel_initializer='he_normal', activation='relu'))
classifier.add(Dense(units=1, kernel_initializer='uniform', activation='sigmoid'))

/usr/local/lib/python3.12/dist-packages/keras/src/layers/core/dense.py:106: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


In [180]:
#compiling the ANN
classifier.compile(optimizer='adam',loss='binary_crossentropy',metrics=['accuracy'])
y_train.iloc[0]

np.int64(0)

In [181]:
classifier.fit(x_train, y_train, batch_size=10,epochs=100)

Epoch 1/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 3s 2ms/step - accuracy: 0.7943 - loss: 0.4642
Epoch 2/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 3s 2ms/step - accuracy: 0.8145 - loss: 0.3992
Epoch 3/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 3s 2ms/step - accuracy: 0.8486 - loss: 0.3782
Epoch 4/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - accuracy: 0.8494 - loss: 0.3682
Epoch 5/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8526 - loss: 0.3631
Epoch 6/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8534 - loss: 0.3609
Epoch 7/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8528 - loss: 0.3590
Epoch 8/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8536 - loss: 0.3572
Epoch 9/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8534 - loss: 0.3575
Epoch 10/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8568 - loss: 0.3553
Epoch 11/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - accuracy: 0.8537 - loss: 0.3537
Epoch 12/100
800/800 ━━━━━━━━━━━━━━━━━━━━

In [182]:
y_pred_ann=classifier.predict(x_test)
y_pred_ann=(y_pred_ann>0.5)
cm3=confusion_matrix(y_test,y_pred_ann)
print(cm3)

63/63 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step
[[1522   73]
 [ 203  202]]


In [183]:
from sklearn.metrics import accuracy_score
print(accuracy_score(y_test,y_pred_ann))

0.862


In [184]:
#Initializing ANN
#kernel_initializer = glorot_uniform

classifier = Sequential()

classifier.add(Dense(units=6, kernel_initializer='uniform', activation='relu', input_dim=10))
classifier.add(Dense(units=6, kernel_initializer='glorot_uniform', activation='relu'))
classifier.add(Dense(units=6, kernel_initializer='glorot_uniform', activation='relu'))
classifier.add(Dense(units=1, kernel_initializer='uniform', activation='sigmoid'))

/usr/local/lib/python3.12/dist-packages/keras/src/layers/core/dense.py:106: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


In [185]:
#compiling the ANN
classifier.compile(optimizer='adam',loss='binary_crossentropy',metrics=['accuracy'])
y_train.iloc[0]

np.int64(0)

In [186]:
classifier.fit(x_train, y_train, batch_size=10,epochs=100)

Epoch 1/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 3s 2ms/step - accuracy: 0.7965 - loss: 0.4822
Epoch 2/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.8264 - loss: 0.4109
Epoch 3/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.8351 - loss: 0.3901
Epoch 4/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8424 - loss: 0.3762
Epoch 5/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8503 - loss: 0.3651
Epoch 6/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8525 - loss: 0.3588
Epoch 7/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8519 - loss: 0.3548
Epoch 8/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.8572 - loss: 0.3520
Epoch 9/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8577 - loss: 0.3503
Epoch 10/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.8560 - loss: 0.3486
Epoch 11/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.8550 - loss: 0.3498
Epoch 12/100
800/800 ━━━━━━━━━━━━━━━━━━━━

In [187]:
y_pred_ann=classifier.predict(x_test)
y_pred_ann=(y_pred_ann>0.5)
cm3=confusion_matrix(y_test,y_pred_ann)
print(cm3)

63/63 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step
[[1509   86]
 [ 199  206]]


In [188]:
from sklearn.metrics import accuracy_score
print(accuracy_score(y_test,y_pred_ann))

0.8575


In [189]:
#Initializing ANN
#kernel_initializer = glorot_normal

classifier = Sequential()

classifier.add(Dense(units=6, kernel_initializer='uniform', activation='relu', input_dim=10))
classifier.add(Dense(units=6, kernel_initializer='glorot_normal', activation='relu'))
classifier.add(Dense(units=6, kernel_initializer='glorot_normal', activation='relu'))
classifier.add(Dense(units=1, kernel_initializer='uniform', activation='sigmoid'))

/usr/local/lib/python3.12/dist-packages/keras/src/layers/core/dense.py:106: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


In [190]:
#compiling the ANN
classifier.compile(optimizer='adam',loss='binary_crossentropy',metrics=['accuracy'])
y_train.iloc[0]

np.int64(0)

In [191]:
classifier.fit(x_train, y_train, batch_size=10,epochs=100)

Epoch 1/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 3s 2ms/step - accuracy: 0.8019 - loss: 0.4626
Epoch 2/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8290 - loss: 0.3931
Epoch 3/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 3s 2ms/step - accuracy: 0.8485 - loss: 0.3751
Epoch 4/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8510 - loss: 0.3660
Epoch 5/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - accuracy: 0.8536 - loss: 0.3625
Epoch 6/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8556 - loss: 0.3579
Epoch 7/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8575 - loss: 0.3540
Epoch 8/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8586 - loss: 0.3523
Epoch 9/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8583 - loss: 0.3498
Epoch 10/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8558 - loss: 0.3498
Epoch 11/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.8581 - loss: 0.3482
Epoch 12/100
800/800 ━━━━━━━━━━━━━━━━━━━━

In [192]:
y_pred_ann=classifier.predict(x_test)
y_pred_ann=(y_pred_ann>0.5)
cm3=confusion_matrix(y_test,y_pred_ann)
print(cm3)

63/63 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step
[[1519   76]
 [ 200  205]]


In [193]:
from sklearn.metrics import accuracy_score
print(accuracy_score(y_test,y_pred_ann))

0.862


In [194]:
#Initializing ANN
#kernel_initializer = random_normal

classifier = Sequential()

classifier.add(Dense(units=6, kernel_initializer='uniform', activation='relu', input_dim=10))
classifier.add(Dense(units=6, kernel_initializer='random_normal', activation='relu'))
classifier.add(Dense(units=6, kernel_initializer='random_normal', activation='relu'))
classifier.add(Dense(units=1, kernel_initializer='uniform', activation='sigmoid'))

/usr/local/lib/python3.12/dist-packages/keras/src/layers/core/dense.py:106: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


In [195]:
#compiling the ANN
classifier.compile(optimizer='adam',loss='binary_crossentropy',metrics=['accuracy'])
y_train.iloc[0]

np.int64(0)

In [196]:
classifier.fit(x_train, y_train, batch_size=10,epochs=100)

Epoch 1/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 4s 2ms/step - accuracy: 0.7958 - loss: 0.4904
Epoch 2/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.7960 - loss: 0.4341
Epoch 3/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.7964 - loss: 0.4279
Epoch 4/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8223 - loss: 0.4230
Epoch 5/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8265 - loss: 0.4207
Epoch 6/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8285 - loss: 0.4196
Epoch 7/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 3s 3ms/step - accuracy: 0.8315 - loss: 0.4178
Epoch 8/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8322 - loss: 0.4163
Epoch 9/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8324 - loss: 0.4148
Epoch 10/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8324 - loss: 0.4135
Epoch 11/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8329 - loss: 0.4120
Epoch 12/100
800/800 ━━━━━━━━━━━━━━━━━━━━

In [197]:
y_pred_ann=classifier.predict(x_test)
y_pred_ann=(y_pred_ann>0.5)
cm3=confusion_matrix(y_test,y_pred_ann)
print(cm3)

63/63 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step
[[1546   49]
 [ 272  133]]


In [198]:
from sklearn.metrics import accuracy_score
print(accuracy_score(y_test,y_pred_ann))

0.8395


Compare

Initial Loss
Final Accuracy
Training Speed

### ANN Model Comparison by Kernel Initializer

| Kernel Initializer | Initial Epoch Training Loss | Final Test Accuracy | Training Speed (approx. total time for 100 epochs) |
|--------------------|-----------------------------|---------------------|--------------------------------------------------|
| `he_uniform`       | 0.4782                      | 0.864               | (Not explicitly captured, requires re-run)       |
| `he_normal`        | 0.4726                      | 0.8565              | (Not explicitly captured, requires re-run)       |
| `glorot_uniform`   | 0.4732                      | 0.8595              | (Not explicitly captured, requires re-run)       |
| `random_normal`    | 0.4913                      | 0.863               | (Not explicitly captured, requires re-run)       |

In [199]:
#TASK - 4 COMPLETE

In [200]:
# TASK 5 BEGIN

In [212]:
#Initializing ANN
# activation : 'sigmoid'

classifier = Sequential()

classifier.add(Dense(units=6, kernel_initializer='uniform', activation='relu', input_dim=10))
classifier.add(Dense(units=6, kernel_initializer='uniform', activation='relu'))
classifier.add(Dense(units=6, kernel_initializer='uniform', activation='relu'))
classifier.add(Dense(units=1, kernel_initializer='uniform', activation='sigmoid'))

/usr/local/lib/python3.12/dist-packages/keras/src/layers/core/dense.py:106: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


In [213]:
#compiling the ANN
classifier.compile(optimizer='adam',loss='binary_crossentropy',metrics=['accuracy'])
y_train.iloc[0]

np.int64(0)

In [214]:
classifier.fit(x_train, y_train, batch_size=10, epochs=100)

Epoch 1/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 3s 2ms/step - accuracy: 0.7950 - loss: 0.4962
Epoch 2/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.7960 - loss: 0.4359
Epoch 3/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.7960 - loss: 0.4324
Epoch 4/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 3s 2ms/step - accuracy: 0.7960 - loss: 0.4283
Epoch 5/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8152 - loss: 0.4239
Epoch 6/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.8257 - loss: 0.4214
Epoch 7/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8299 - loss: 0.4190
Epoch 8/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8301 - loss: 0.4183
Epoch 9/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8316 - loss: 0.4167
Epoch 10/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8314 - loss: 0.4159
Epoch 11/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 3s 2ms/step - accuracy: 0.8326 - loss: 0.4145
Epoch 12/100
800/800 ━━━━━━━━━━━━━━━━━━━━

In [215]:
y_pred_ann=classifier.predict(x_test)
y_pred_ann=(y_pred_ann>0.5)
cm3=confusion_matrix(y_test,y_pred_ann)
print(cm3)

63/63 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step
[[1542   53]
 [ 263  142]]


In [216]:
from sklearn.metrics import accuracy_score
print(accuracy_score(y_test,y_pred_ann))

0.842


In [217]:
#Initializing ANN
#activation : 'softmax'

classifier = Sequential()

classifier.add(Dense(units=6, kernel_initializer='uniform', activation='relu', input_dim=10))
classifier.add(Dense(units=6, kernel_initializer='uniform', activation='relu'))
classifier.add(Dense(units=6, kernel_initializer='uniform', activation='relu'))
classifier.add(Dense(units=1, kernel_initializer='uniform', activation='softmax'))

/usr/local/lib/python3.12/dist-packages/keras/src/layers/core/dense.py:106: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


In [218]:
#compiling the ANN
classifier.compile(optimizer='adam',loss='binary_crossentropy',metrics=['accuracy'])
y_train.iloc[0]

np.int64(0)

In [219]:
classifier.fit(x_train, y_train, batch_size=10, epochs=100)

Epoch 1/100


/usr/local/lib/python3.12/dist-packages/keras/src/ops/nn.py:947: UserWarning: You are using a softmax over axis -1 of a tensor of shape (10, 1). This axis has size 1. The softmax operation will always return the value 1, which is likely not what you intended. Did you mean to use a sigmoid instead?
  warnings.warn(


800/800 ━━━━━━━━━━━━━━━━━━━━ 3s 2ms/step - accuracy: 0.2040 - loss: 0.4958
Epoch 2/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.2040 - loss: 0.4294
Epoch 3/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.2040 - loss: 0.4208
Epoch 4/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.2040 - loss: 0.4064
Epoch 5/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - accuracy: 0.2040 - loss: 0.3878
Epoch 6/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.2040 - loss: 0.3754
Epoch 7/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.2040 - loss: 0.3689
Epoch 8/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.2040 - loss: 0.3649
Epoch 9/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.2040 - loss: 0.3619
Epoch 10/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.2040 - loss: 0.3596
Epoch 11/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.2040 - loss: 0.3571
Epoch 12/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step

In [220]:
y_pred_ann=classifier.predict(x_test)
y_pred_ann=(y_pred_ann>0.5)
cm3=confusion_matrix(y_test,y_pred_ann)
print(cm3)

63/63 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step
[[   0 1595]
 [   0  405]]


/usr/local/lib/python3.12/dist-packages/keras/src/ops/nn.py:947: UserWarning: You are using a softmax over axis -1 of a tensor of shape (32, 1). This axis has size 1. The softmax operation will always return the value 1, which is likely not what you intended. Did you mean to use a sigmoid instead?
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/keras/src/ops/nn.py:947: UserWarning: You are using a softmax over axis -1 of a tensor of shape (None, 1). This axis has size 1. The softmax operation will always return the value 1, which is likely not what you intended. Did you mean to use a sigmoid instead?
  warnings.warn(


In [221]:
from sklearn.metrics import accuracy_score
print(accuracy_score(y_test,y_pred_ann))

0.2025


In [222]:
#Initializing ANN
# activation : 'liner

classifier = Sequential()

classifier.add(Dense(units=6, kernel_initializer='uniform', activation='relu', input_dim=10))
classifier.add(Dense(units=6, kernel_initializer='uniform', activation='relu'))
classifier.add(Dense(units=6, kernel_initializer='uniform', activation='relu'))
classifier.add(Dense(units=1, kernel_initializer='uniform', activation='linear'))


/usr/local/lib/python3.12/dist-packages/keras/src/layers/core/dense.py:106: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


In [223]:
#compiling the ANN
classifier.compile(optimizer='adam',loss='binary_crossentropy',metrics=['accuracy'])
y_train.iloc[0]

np.int64(0)

In [224]:
classifier.fit(x_train, y_train, batch_size=10, epochs=100)

Epoch 1/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 4s 2ms/step - accuracy: 0.7960 - loss: 3.2881
Epoch 2/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - accuracy: 0.7960 - loss: 3.2881
Epoch 3/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.7960 - loss: 3.2881
Epoch 4/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.7960 - loss: 3.2881
Epoch 5/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.7960 - loss: 3.2881
Epoch 6/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 3s 4ms/step - accuracy: 0.7960 - loss: 3.2881
Epoch 7/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 7s 6ms/step - accuracy: 0.7960 - loss: 3.2881
Epoch 8/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 3s 4ms/step - accuracy: 0.7960 - loss: 3.2881
Epoch 9/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - accuracy: 0.7960 - loss: 3.2881
Epoch 10/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.7960 - loss: 3.2881
Epoch 11/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 3s 2ms/step - accuracy: 0.7960 - loss: 3.2881
Epoch 12/100
800/800 ━━━━━━━━━━━━━━━━━━━━

In [225]:
y_pred_ann=classifier.predict(x_test)
y_pred_ann=(y_pred_ann>0.5)
cm3=confusion_matrix(y_test,y_pred_ann)
print(cm3)

63/63 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step
[[1595    0]
 [ 405    0]]


In [226]:
from sklearn.metrics import accuracy_score
print(accuracy_score(y_test,y_pred_ann))

0.7975


In [227]:
# TASK 5 COMPLETED

In [228]:
# TASK 6 BEGIN

In [229]:
#Initializing ANN
# optimizer = adam

classifier = Sequential()

classifier.add(Dense(units=6,kernel_initializer='uniform',activation='relu',input_dim=10))
classifier.add(Dense(units=6,kernel_initializer='uniform',activation='relu'))
classifier.add(Dense(units=6,kernel_initializer='uniform',activation='relu'))
classifier.add(Dense(units=1,kernel_initializer='uniform',activation='sigmoid'))

#Compiling the ANN
classifier.compile(optimizer='adam',loss='binary_crossentropy',metrics=['accuracy'])


classifier.fit(x_train,y_train,batch_size=10,epochs=50)

#Prediction
y_pred_ann=classifier.predict(x_test)
y_pred_ann=(y_pred_ann>0.5)
cm3=confusion_matrix(y_test,y_pred_ann)
print(cm3)

from sklearn.metrics import accuracy_score
print(accuracy_score(y_test,y_pred_ann))

Epoch 1/50


/usr/local/lib/python3.12/dist-packages/keras/src/layers/core/dense.py:106: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


800/800 ━━━━━━━━━━━━━━━━━━━━ 4s 3ms/step - accuracy: 0.7954 - loss: 0.5019
Epoch 2/50
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - accuracy: 0.7960 - loss: 0.4352
Epoch 3/50
800/800 ━━━━━━━━━━━━━━━━━━━━ 3s 4ms/step - accuracy: 0.7960 - loss: 0.4307
Epoch 4/50
800/800 ━━━━━━━━━━━━━━━━━━━━ 4s 2ms/step - accuracy: 0.7960 - loss: 0.4265
Epoch 5/50
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8186 - loss: 0.4225
Epoch 6/50
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8257 - loss: 0.4205
Epoch 7/50
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8288 - loss: 0.4184
Epoch 8/50
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8294 - loss: 0.4171
Epoch 9/50
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - accuracy: 0.8305 - loss: 0.4156
Epoch 10/50
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8324 - loss: 0.4140
Epoch 11/50
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8328 - loss: 0.4131
Epoch 12/50
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy

In [233]:
#Initializing ANN
#using optimizer = sdg
classifier = Sequential()

classifier.add(Dense(units=6,kernel_initializer='uniform',activation='relu',input_dim=10))
classifier.add(Dense(units=6,kernel_initializer='uniform',activation='relu'))
classifier.add(Dense(units=6,kernel_initializer='uniform',activation='relu'))
classifier.add(Dense(units=1,kernel_initializer='uniform',activation='sigmoid'))

In [231]:
classifier.compile(optimizer='sgd',loss='binary_crossentropy',metrics=['accuracy'])

In [232]:
classifier.fit(x_train,y_train,batch_size=10,epochs=50)

Epoch 1/50
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.7960 - loss: 0.5572
Epoch 2/50
800/800 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.7960 - loss: 0.5080
Epoch 3/50
800/800 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.7960 - loss: 0.5061
Epoch 4/50
800/800 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.7960 - loss: 0.5060
Epoch 5/50
800/800 ━━━━━━━━━━━━━━━━━━━━ 4s 3ms/step - accuracy: 0.7960 - loss: 0.5059
Epoch 6/50
800/800 ━━━━━━━━━━━━━━━━━━━━ 3s 4ms/step - accuracy: 0.7960 - loss: 0.5060
Epoch 7/50
800/800 ━━━━━━━━━━━━━━━━━━━━ 4s 3ms/step - accuracy: 0.7960 - loss: 0.5060
Epoch 8/50
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - accuracy: 0.7960 - loss: 0.5060
Epoch 9/50
800/800 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.7960 - loss: 0.5060
Epoch 10/50
800/800 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.7960 - loss: 0.5060
Epoch 11/50
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.7960 - loss: 0.5060
Epoch 12/50
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step

In [234]:
y_pred_ann=classifier.predict(x_test)
y_pred_ann=(y_pred_ann>0.5)
cm3=confusion_matrix(y_test,y_pred_ann)
print(cm3)

from sklearn.metrics import accuracy_score
print(accuracy_score(y_test,y_pred_ann))

63/63 ━━━━━━━━━━━━━━━━━━━━ 1s 5ms/step
[[1119  476]
 [ 329   76]]
0.5975


In [235]:
# Initializing ANN
# optimizer = rmsprop

classifier = Sequential()

classifier.add(Dense(units=6,kernel_initializer='uniform',activation='relu',input_dim=10))
classifier.add(Dense(units=6,kernel_initializer='uniform',activation='relu'))
classifier.add(Dense(units=6,kernel_initializer='uniform',activation='relu'))
classifier.add(Dense(units=1,kernel_initializer='uniform',activation='sigmoid'))

/usr/local/lib/python3.12/dist-packages/keras/src/layers/core/dense.py:106: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


In [236]:
classifier.compile(optimizer='rmsprop',loss='binary_crossentropy',metrics=['accuracy'])

In [237]:
classifier.fit(x_train,y_train,batch_size=10,epochs=50)

Epoch 1/50
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.7960 - loss: 0.5152
Epoch 2/50
800/800 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.7960 - loss: 0.4393
Epoch 3/50
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - accuracy: 0.7960 - loss: 0.4296
Epoch 4/50
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8173 - loss: 0.4239
Epoch 5/50
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8290 - loss: 0.4204
Epoch 6/50
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - accuracy: 0.8310 - loss: 0.4187
Epoch 7/50
800/800 ━━━━━━━━━━━━━━━━━━━━ 3s 3ms/step - accuracy: 0.8315 - loss: 0.4174
Epoch 8/50
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - accuracy: 0.8324 - loss: 0.4160
Epoch 9/50
800/800 ━━━━━━━━━━━━━━━━━━━━ 3s 4ms/step - accuracy: 0.8345 - loss: 0.4144
Epoch 10/50
800/800 ━━━━━━━━━━━━━━━━━━━━ 4s 3ms/step - accuracy: 0.8331 - loss: 0.4145
Epoch 11/50
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8336 - loss: 0.4128
Epoch 12/50
800/800 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step

In [239]:
y_pred_ann=classifier.predict(x_test)
y_pred_ann=(y_pred_ann>0.5)

cm3=confusion_matrix(y_test,y_pred_ann)
print(cm3)

from sklearn.metrics import accuracy_score
print(accuracy_score(y_test,y_pred_ann))

63/63 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step
[[1526   69]
 [ 243  162]]
0.844


In [248]:
# Initializing ANN
# optimizer = adagrad
from keras.optimizers import Adagrad

classifier = Sequential()

classifier.add(Dense(units=6,kernel_initializer='uniform',activation='relu',input_dim=10))
classifier.add(Dense(units=6,kernel_initializer='uniform',activation='relu'))
classifier.add(Dense(units=6,kernel_initializer='uniform',activation='relu'))
classifier.add(Dense(units=1,kernel_initializer='uniform',activation='sigmoid'))

/usr/local/lib/python3.12/dist-packages/keras/src/layers/core/dense.py:106: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


In [249]:
classifier.compile(optimizer=Adagrad(learning_rate=0.01),loss='binary_crossentropy',metrics=['accuracy'])

In [250]:
classifier.fit(x_train,y_train,batch_size=10,epochs=100)

Epoch 1/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.7958 - loss: 0.6051
Epoch 2/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.7960 - loss: 0.5337
Epoch 3/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.7960 - loss: 0.5120
Epoch 4/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.7960 - loss: 0.5069
Epoch 5/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.7960 - loss: 0.5051
Epoch 6/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.7960 - loss: 0.5018
Epoch 7/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.7960 - loss: 0.4920
Epoch 8/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - accuracy: 0.7960 - loss: 0.4747
Epoch 9/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.7960 - loss: 0.4618
Epoch 10/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 3s 2ms/step - accuracy: 0.7960 - loss: 0.4471
Epoch 11/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.7960 - loss: 0.4351
Epoch 12/100
800/800 ━━━━━━━━━━━━━━━━━━━━

In [255]:
y_pred_ann=classifier.predict(x_test)
y_pred_ann=(y_pred_ann>0.5)
cm3=confusion_matrix(y_test,y_pred_ann)
print(cm3)

from sklearn.metrics import accuracy_score
print(accuracy_score(y_test,y_pred_ann))

63/63 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step
[[1551   44]
 [ 271  134]]
0.8425


In [265]:
# Initializing ANN and compiling for Nadam
from keras.optimizers import Nadam

classifier = Sequential()

classifier.add(Dense(units=6,kernel_initializer='uniform',activation='relu',input_dim=10))
classifier.add(Dense(units=6,kernel_initializer='uniform',activation='relu'))
classifier.add(Dense(units=6,kernel_initializer='uniform',activation='relu'))
classifier.add(Dense(units=1,kernel_initializer='uniform',activation='sigmoid'))

classifier.compile(optimizer=Nadam(learning_rate=0.01),loss='binary_crossentropy',metrics=['accuracy'])

/usr/local/lib/python3.12/dist-packages/keras/src/layers/core/dense.py:106: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


In [266]:
# Nadam import (compilation moved to model initialization cell)

In [267]:
classifier.fit(x_train,y_train,batch_size=10,epochs=100)

Epoch 1/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 4s 3ms/step - accuracy: 0.7960 - loss: 0.4502
Epoch 2/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8248 - loss: 0.4298
Epoch 3/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8300 - loss: 0.4253
Epoch 4/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8314 - loss: 0.4248
Epoch 5/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8329 - loss: 0.4203
Epoch 6/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8321 - loss: 0.4206
Epoch 7/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 3s 2ms/step - accuracy: 0.8314 - loss: 0.4189
Epoch 8/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 3s 3ms/step - accuracy: 0.8314 - loss: 0.4160
Epoch 9/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8324 - loss: 0.4175
Epoch 10/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 3s 3ms/step - accuracy: 0.8346 - loss: 0.4144
Epoch 11/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8316 - loss: 0.4139
Epoch 12/100
800/800 ━━━━━━━━━━━━━━━━━━━━

In [270]:
y_pred_ann=classifier.predict(x_test)
y_pred_ann=(y_pred_ann>0.5)
cm3=confusion_matrix(y_test,y_pred_ann)
print(cm3)

from sklearn.metrics import accuracy_score
print(accuracy_score(y_test,y_pred_ann))

63/63 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step
[[1530   65]
 [ 213  192]]
0.861


create a comparision report in text field, with respect to the optimiers used

### ANN Optimizer Comparison

| Optimizer         | Final Test Accuracy | Last Epoch Training Loss |
|-------------------|---------------------|--------------------------|
| Adam              | 0.8445              | 0.3981                   |
| SGD               | 0.5975              | 0.5060                   |
| RMSprop           | 0.8440              | 0.4016                   |
| Adagrad           | 0.8425              | 0.3524                   |
| Nadam             | 0.861               | 0.3495                   |

In [271]:
# TASK 6 COMPLETED

In [272]:
# TASK 7 BEGIN

In [273]:
#Initializing ANN
# EPOCHS = 10

classifier = Sequential()

classifier.add(Dense(units=6,kernel_initializer='uniform',activation='relu',input_dim=10))
classifier.add(Dense(units=6,kernel_initializer='uniform',activation='relu'))
classifier.add(Dense(units=6,kernel_initializer='uniform',activation='relu'))
classifier.add(Dense(units=1,kernel_initializer='uniform',activation='sigmoid'))

#Compiling the ANN
classifier.compile(optimizer='adam',loss='binary_crossentropy',metrics=['accuracy'])

classifier.fit(x_train,y_train,batch_size=10,epochs=10)

y_pred_ann=classifier.predict(x_test)
y_pred_ann=(y_pred_ann>0.5)
cm3=confusion_matrix(y_test,y_pred_ann)
print(cm3)

from sklearn.metrics import accuracy_score
print(accuracy_score(y_test,y_pred_ann))

Epoch 1/10


/usr/local/lib/python3.12/dist-packages/keras/src/layers/core/dense.py:106: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


800/800 ━━━━━━━━━━━━━━━━━━━━ 5s 3ms/step - accuracy: 0.7951 - loss: 0.6146
Epoch 2/10
800/800 ━━━━━━━━━━━━━━━━━━━━ 4s 4ms/step - accuracy: 0.7960 - loss: 0.5347
Epoch 3/10
800/800 ━━━━━━━━━━━━━━━━━━━━ 3s 2ms/step - accuracy: 0.7960 - loss: 0.5122
Epoch 4/10
800/800 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.7960 - loss: 0.5070
Epoch 5/10
800/800 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.7960 - loss: 0.5061
Epoch 6/10
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - accuracy: 0.7960 - loss: 0.5060
Epoch 7/10
800/800 ━━━━━━━━━━━━━━━━━━━━ 4s 5ms/step - accuracy: 0.7960 - loss: 0.5059
Epoch 8/10
800/800 ━━━━━━━━━━━━━━━━━━━━ 4s 3ms/step - accuracy: 0.7960 - loss: 0.5060
Epoch 9/10
800/800 ━━━━━━━━━━━━━━━━━━━━ 3s 3ms/step - accuracy: 0.7960 - loss: 0.5060
Epoch 10/10
800/800 ━━━━━━━━━━━━━━━━━━━━ 5s 4ms/step - accuracy: 0.7960 - loss: 0.5059
63/63 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step
[[1595    0]
 [ 405    0]]
0.7975


In [274]:
#Initializing ANN
# EPOCHS = 25

classifier = Sequential()

classifier.add(Dense(units=6,kernel_initializer='uniform',activation='relu',input_dim=10))
classifier.add(Dense(units=6,kernel_initializer='uniform',activation='relu'))
classifier.add(Dense(units=6,kernel_initializer='uniform',activation='relu'))
classifier.add(Dense(units=1,kernel_initializer='uniform',activation='sigmoid'))

#Compiling the ANN
classifier.compile(optimizer='adam',loss='binary_crossentropy',metrics=['accuracy'])

classifier.fit(x_train,y_train,batch_size=10,epochs=25)

y_pred_ann=classifier.predict(x_test)
y_pred_ann=(y_pred_ann>0.5)

cm3=confusion_matrix(y_test,y_pred_ann)
print(cm3)

from sklearn.metrics import accuracy_score
print(accuracy_score(y_test,y_pred_ann))

Epoch 1/25


/usr/local/lib/python3.12/dist-packages/keras/src/layers/core/dense.py:106: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


800/800 ━━━━━━━━━━━━━━━━━━━━ 5s 3ms/step - accuracy: 0.7958 - loss: 0.4811
Epoch 2/25
800/800 ━━━━━━━━━━━━━━━━━━━━ 7s 6ms/step - accuracy: 0.7960 - loss: 0.4341
Epoch 3/25
800/800 ━━━━━━━━━━━━━━━━━━━━ 3s 4ms/step - accuracy: 0.7985 - loss: 0.4272
Epoch 4/25
800/800 ━━━━━━━━━━━━━━━━━━━━ 3s 4ms/step - accuracy: 0.8225 - loss: 0.4225
Epoch 5/25
800/800 ━━━━━━━━━━━━━━━━━━━━ 4s 2ms/step - accuracy: 0.8271 - loss: 0.4196
Epoch 6/25
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - accuracy: 0.8285 - loss: 0.4166
Epoch 7/25
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8313 - loss: 0.4153
Epoch 8/25
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8307 - loss: 0.4139
Epoch 9/25
800/800 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.8328 - loss: 0.4122
Epoch 10/25
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8330 - loss: 0.4114
Epoch 11/25
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8339 - loss: 0.4108
Epoch 12/25
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy

In [276]:
#Initializing ANN
# EPOCHS = 50

classifier = Sequential()

classifier.add(Dense(units=6,kernel_initializer='uniform',activation='relu',input_dim=10))
classifier.add(Dense(units=6,kernel_initializer='uniform',activation='relu'))
classifier.add(Dense(units=6,kernel_initializer='uniform',activation='relu'))
classifier.add(Dense(units=1,kernel_initializer='uniform',activation='sigmoid'))

#Compiling the ANN
classifier.compile(optimizer='adam',loss='binary_crossentropy',metrics=['accuracy'])

classifier.fit(x_train,y_train,batch_size=10,epochs=50)

y_pred_ann=classifier.predict(x_test)
y_pred_ann=(y_pred_ann>0.5)
cm3=confusion_matrix(y_test,y_pred_ann)
print(cm3)

from sklearn.metrics import accuracy_score
print(accuracy_score(y_test,y_pred_ann))

Epoch 1/50
800/800 ━━━━━━━━━━━━━━━━━━━━ 5s 4ms/step - accuracy: 0.7956 - loss: 0.4866
Epoch 2/50
800/800 ━━━━━━━━━━━━━━━━━━━━ 4s 5ms/step - accuracy: 0.7960 - loss: 0.4334
Epoch 3/50
800/800 ━━━━━━━━━━━━━━━━━━━━ 3s 2ms/step - accuracy: 0.7962 - loss: 0.4273
Epoch 4/50
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8202 - loss: 0.4226
Epoch 5/50
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8288 - loss: 0.4199
Epoch 6/50
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8310 - loss: 0.4172
Epoch 7/50
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8313 - loss: 0.4152
Epoch 8/50
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8313 - loss: 0.4130
Epoch 9/50
800/800 ━━━━━━━━━━━━━━━━━━━━ 3s 2ms/step - accuracy: 0.8328 - loss: 0.4120
Epoch 10/50
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8346 - loss: 0.4103
Epoch 11/50
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8326 - loss: 0.4096
Epoch 12/50
800/800 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step

In [277]:
#Initializing ANN
# EPOCHS = 100

classifier = Sequential()

classifier.add(Dense(units=6,kernel_initializer='uniform',activation='relu',input_dim=10))
classifier.add(Dense(units=6,kernel_initializer='uniform',activation='relu'))
classifier.add(Dense(units=6,kernel_initializer='uniform',activation='relu'))
classifier.add(Dense(units=1,kernel_initializer='uniform',activation='sigmoid'))

#Compiling the ANN
classifier.compile(optimizer='adam',loss='binary_crossentropy',metrics=['accuracy'])

classifier.fit(x_train,y_train,batch_size=10,epochs=100)

y_pred_ann=classifier.predict(x_test)
y_pred_ann=(y_pred_ann>0.5)
cm3=confusion_matrix(y_test,y_pred_ann)
print(cm3)

from sklearn.metrics import accuracy_score
print(accuracy_score(y_test,y_pred_ann))

Epoch 1/100


/usr/local/lib/python3.12/dist-packages/keras/src/layers/core/dense.py:106: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


800/800 ━━━━━━━━━━━━━━━━━━━━ 7s 5ms/step - accuracy: 0.7955 - loss: 0.4885
Epoch 2/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 4s 5ms/step - accuracy: 0.7960 - loss: 0.4362
Epoch 3/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.7960 - loss: 0.4343
Epoch 4/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.7960 - loss: 0.4325
Epoch 5/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.7960 - loss: 0.4293
Epoch 6/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8129 - loss: 0.4258
Epoch 7/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8221 - loss: 0.4221
Epoch 8/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - accuracy: 0.8256 - loss: 0.4190
Epoch 9/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 3s 3ms/step - accuracy: 0.8270 - loss: 0.4166
Epoch 10/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8289 - loss: 0.4148
Epoch 11/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8313 - loss: 0.4130
Epoch 12/100
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step

In [278]:
#Initializing ANN
# EPOCHS = 200

classifier = Sequential()

classifier.add(Dense(units=6,kernel_initializer='uniform',activation='relu',input_dim=10))
classifier.add(Dense(units=6,kernel_initializer='uniform',activation='relu'))
classifier.add(Dense(units=6,kernel_initializer='uniform',activation='relu'))
classifier.add(Dense(units=1,kernel_initializer='uniform',activation='sigmoid'))

#Compiling the ANN
classifier.compile(optimizer='adam',loss='binary_crossentropy',metrics=['accuracy'])

classifier.fit(x_train,y_train,batch_size=10,epochs=200)

y_pred_ann=classifier.predict(x_test)
y_pred_ann=(y_pred_ann>0.5)
cm3=confusion_matrix(y_test,y_pred_ann)
print(cm3)

from sklearn.metrics import accuracy_score
print(accuracy_score(y_test,y_pred_ann))

/usr/local/lib/python3.12/dist-packages/keras/src/layers/core/dense.py:106: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Epoch 1/200
800/800 ━━━━━━━━━━━━━━━━━━━━ 6s 3ms/step - accuracy: 0.7958 - loss: 0.4983
Epoch 2/200
800/800 ━━━━━━━━━━━━━━━━━━━━ 3s 3ms/step - accuracy: 0.7960 - loss: 0.4337
Epoch 3/200
800/800 ━━━━━━━━━━━━━━━━━━━━ 4s 2ms/step - accuracy: 0.7960 - loss: 0.4289
Epoch 4/200
800/800 ━━━━━━━━━━━━━━━━━━━━ 3s 3ms/step - accuracy: 0.8069 - loss: 0.4251
Epoch 5/200
800/800 ━━━━━━━━━━━━━━━━━━━━ 4s 2ms/step - accuracy: 0.8220 - loss: 0.4222
Epoch 6/200
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8261 - loss: 0.4204
Epoch 7/200
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8306 - loss: 0.4186
Epoch 8/200
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8320 - loss: 0.4174
Epoch 9/200
800/800 ━━━━━━━━━━━━━━━━━━━━ 3s 3ms/step - accuracy: 0.8313 - loss: 0.4163
Epoch 10/200
800/800 ━━━━━━━━━━━━━━━━━━━━ 4s 2ms/step - accuracy: 0.8315 - loss: 0.4158
Epoch 11/200
800/800 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8332 - loss: 0.4144
Epoch 12/200
800/800 ━━━━━━━━━━━━━━━━━━━━

In [279]:
# TASK 7 COMPLETED

In [280]:
#Initializing ANN


classifier = Sequential()

classifier.add(Dense(units=6,kernel_initializer='uniform',activation='relu',input_dim=10))
classifier.add(Dense(units=6,kernel_initializer='uniform',activation='relu'))
classifier.add(Dense(units=6,kernel_initializer='uniform',activation='relu'))
classifier.add(Dense(units=1,kernel_initializer='uniform',activation='sigmoid'))

#Compiling the ANN
classifier.compile(optimizer='adam',loss='binary_crossentropy',metrics=['accuracy'])

classifier.fit(x_train,y_train,batch_size=8,epochs=10)

y_pred_ann=classifier.predict(x_test)
y_pred_ann=(y_pred_ann>0.5)

cm3=confusion_matrix(y_test,y_pred_ann)
print(cm3)

from sklearn.metrics import accuracy_score
print(accuracy_score(y_test,y_pred_ann))

Epoch 1/10


/usr/local/lib/python3.12/dist-packages/keras/src/layers/core/dense.py:106: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


1000/1000 ━━━━━━━━━━━━━━━━━━━━ 7s 5ms/step - accuracy: 0.7956 - loss: 0.4755
Epoch 2/10
1000/1000 ━━━━━━━━━━━━━━━━━━━━ 8s 3ms/step - accuracy: 0.7960 - loss: 0.4313
Epoch 3/10
1000/1000 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.8114 - loss: 0.4243
Epoch 4/10
1000/1000 ━━━━━━━━━━━━━━━━━━━━ 5s 5ms/step - accuracy: 0.8250 - loss: 0.4214
Epoch 5/10
1000/1000 ━━━━━━━━━━━━━━━━━━━━ 4s 4ms/step - accuracy: 0.8275 - loss: 0.4195
Epoch 6/10
1000/1000 ━━━━━━━━━━━━━━━━━━━━ 3s 3ms/step - accuracy: 0.8292 - loss: 0.4176
Epoch 7/10
1000/1000 ━━━━━━━━━━━━━━━━━━━━ 4s 4ms/step - accuracy: 0.8311 - loss: 0.4166
Epoch 8/10
1000/1000 ━━━━━━━━━━━━━━━━━━━━ 6s 5ms/step - accuracy: 0.8321 - loss: 0.4153
Epoch 9/10
1000/1000 ━━━━━━━━━━━━━━━━━━━━ 5s 5ms/step - accuracy: 0.8330 - loss: 0.4139
Epoch 10/10
1000/1000 ━━━━━━━━━━━━━━━━━━━━ 7s 2ms/step - accuracy: 0.8320 - loss: 0.4133
63/63 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step
[[1554   41]
 [ 288  117]]
0.8355


In [281]:
#Initializing ANN

classifier = Sequential()

classifier.add(Dense(units=6,kernel_initializer='uniform',activation='relu',input_dim=10))
classifier.add(Dense(units=6,kernel_initializer='uniform',activation='relu'))
classifier.add(Dense(units=6,kernel_initializer='uniform',activation='relu'))
classifier.add(Dense(units=1,kernel_initializer='uniform',activation='sigmoid'))

#Compiling the ANN
classifier.compile(optimizer='adam',loss='binary_crossentropy',metrics=['accuracy'])

classifier.fit(x_train,y_train,batch_size=16,epochs=10)

y_pred_ann=classifier.predict(x_test)
y_pred_ann=(y_pred_ann>0.5)

cm3=confusion_matrix(y_test,y_pred_ann)
print(cm3)

from sklearn.metrics import accuracy_score
print(accuracy_score(y_test,y_pred_ann))

Epoch 1/10


/usr/local/lib/python3.12/dist-packages/keras/src/layers/core/dense.py:106: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


500/500 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.7959 - loss: 0.5151
Epoch 2/10
500/500 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - accuracy: 0.7960 - loss: 0.4361
Epoch 3/10
500/500 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.7960 - loss: 0.4304
Epoch 4/10
500/500 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.7977 - loss: 0.4252
Epoch 5/10
500/500 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.8205 - loss: 0.4217
Epoch 6/10
500/500 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.8286 - loss: 0.4192
Epoch 7/10
500/500 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.8295 - loss: 0.4174
Epoch 8/10
500/500 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.8313 - loss: 0.4156
Epoch 9/10
500/500 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.8328 - loss: 0.4138
Epoch 10/10
500/500 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.8320 - loss: 0.4128
63/63 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step
[[1564   31]
 [ 298  107]]
0.8355


In [ ]:
#Initializing ANN

classifier = Sequential()

classifier.add(Dense(units=6,kernel_initializer='uniform',activation='relu',input_dim=10))
classifier.add(Dense(units=6,kernel_initializer='uniform',activation='relu'))
classifier.add(Dense(units=6,kernel_initializer='uniform',activation='relu'))
classifier.add(Dense(units=1,kernel_initializer='uniform',activation='sigmoid'))

#Compiling the ANN
classifier.compile(optimizer='adam',loss='binary_crossentropy',metrics=['accuracy'])

classifier.fit(x_train,y_train,batch_size=32,epochs=10)

y_pred_ann=classifier.predict(x_test)
y_pred_ann=(y_pred_ann>0.5)

cm3=confusion_matrix(y_test,y_pred_ann)
print(cm3)

from sklearn.metrics import accuracy_score
print(accuracy_score(y_test,y_pred_ann))

In [282]:
# TASK 8 COMPLETED

In [283]:
# TASK 9

### Task 9: Improve the Existing ANN Model

To improve the model's accuracy by at least 5%, we will systematically explore different configurations, including:
-   **Hidden Layers and Neurons:** Further tuning the number of hidden layers and neurons per layer.
-   **Optimizer:** Fine-tuning the learning rate of existing optimizers or exploring new ones.
-   **Activation Functions:** Re-evaluating activation functions in conjunction with other changes.
-   **Dropout:** Adding dropout layers to prevent overfitting.
-   **Epochs:** Adjusting the number of training epochs.

We will start by defining a flexible function to build, compile, train, and evaluate the ANN model, making it easier to experiment with these changes.

In [285]:
from keras.layers import Dropout

def build_and_evaluate_ann(num_hidden_layers, neurons_per_layer, optimizer_name='adam', learning_rate=0.001, activation_func='relu', dropout_rate=0.0, epochs=100, batch_size=10, kernel_initializer='uniform'):
    # Initialize the ANN
    classifier = Sequential()

    # Input layer
    classifier.add(Dense(units=neurons_per_layer, kernel_initializer=kernel_initializer, activation=activation_func, input_dim=x_train.shape[1]))
    if dropout_rate > 0.0:
        classifier.add(Dropout(dropout_rate))

    # Hidden layers
    for _ in range(num_hidden_layers - 1):
        classifier.add(Dense(units=neurons_per_layer, kernel_initializer=kernel_initializer, activation=activation_func))
        if dropout_rate > 0.0:
            classifier.add(Dropout(dropout_rate))

    # Output layer
    classifier.add(Dense(units=1, kernel_initializer='uniform', activation='sigmoid'))

    # Compile the ANN
    optimizer = None
    if optimizer_name == 'adam':
        optimizer = keras.optimizers.Adam(learning_rate=learning_rate)
    elif optimizer_name == 'sgd':
        optimizer = keras.optimizers.SGD(learning_rate=learning_rate)
    elif optimizer_name == 'rmsprop':
        optimizer = keras.optimizers.RMSprop(learning_rate=learning_rate)
    elif optimizer_name == 'adagrad':
        optimizer = keras.optimizers.Adagrad(learning_rate=learning_rate)
    elif optimizer_name == 'nadam':
        optimizer = keras.optimizers.Nadam(learning_rate=learning_rate)
    else:
        raise ValueError(f"Unknown optimizer: {optimizer_name}")

    classifier.compile(optimizer=optimizer, loss='binary_crossentropy', metrics=['accuracy'])

    # Train the ANN
    print(f"\nTraining ANN with {num_hidden_layers} hidden layers, {neurons_per_layer} neurons, {optimizer_name} (lr={learning_rate}), {activation_func} activation, dropout={dropout_rate}, epochs={epochs}, batch_size={batch_size}")
    history = classifier.fit(x_train, y_train, batch_size=batch_size, epochs=epochs, verbose=0)

    # Predict and evaluate
    y_pred_ann = classifier.predict(x_test)
    y_pred_ann = (y_pred_ann > 0.5)
    cm = confusion_matrix(y_test, y_pred_ann)
    accuracy = accuracy_score(y_test, y_pred_ann)
    final_loss = history.history['loss'][-1]

    print(f"Confusion Matrix:\n{cm}")
    print(f"Test Accuracy: {accuracy:.4f}")
    print(f"Last Epoch Training Loss: {final_loss:.4f}")

    return accuracy, final_loss


In [ ]:
# TASK COMPLETED